### ISSUE TITLE
acme_certificate renewal fails with "First certificate is a CA certificate"

### ISSUE THREAD

**Issue by uovobw** [terraform-providers/terraform-provider-acme#54]
The certificate renewal (with latest terraform `0.11.14`)  with the acme provider (latest version `1.3.0`) consistently fails during the `terraform plan` stage with the error:

```
Error: Error running plan: 1 error occurred:
	* acme_certificate.cc_acme_certificate: 1 error occurred:
	* acme_certificate.cc_acme_certificate: unable to calculate time to certificate expiry: First certificate is a CA certificate
```

Below is the configuration that causes the problem:

```
resource "tls_private_key" "private_staging_key" {
  algorithm = "RSA"
  rsa_bits = 4096
  lifecycle {
    create_before_destroy = true
  }
}

resource "tls_private_key" "private_prod_key" {
  algorithm = "RSA"
  rsa_bits = 4096
  lifecycle {
    create_before_destroy = true
  }
}

resource "acme_registration" "cc_staging_registration" {
  account_key_pem = "${tls_private_key.private_staging_key.private_key_pem}"
  email_address   = "some-email@email.com"
  provider = "acme.staging"
  lifecycle {
    create_before_destroy = true
  }
}

resource "acme_registration" "cc_prod_registration" {
  account_key_pem = "${tls_private_key.private_prod_key.private_key_pem}"
  email_address   = "some-email@email.com"
  provider = "acme.prod"
  lifecycle {
    create_before_destroy = true
  }
}

resource "acme_certificate" "cc_acme_certificate" {
  account_key_pem           = "${acme_registration.cc_prod_registration.account_key_pem}"
  common_name               = "*.our-domain.com"
  provider = "acme.prod"
  dns_challenge {
    provider = "route53"
  }
  lifecycle {
    create_before_destroy = true
  }
}

resource "aws_acm_certificate" "cc_ssl_certificate" {
  private_key = "${acme_certificate.cc_acme_certificate.private_key_pem}"
  certificate_body = "${acme_certificate.cc_acme_certificate.certificate_pem}"
  certificate_chain = "${acme_certificate.cc_acme_certificate.issuer_pem}"
}
```

The `acme` provider is configured as follows:

```
provider "acme" {
  alias = "prod"
  server_url = "https://acme-v02.api.letsencrypt.org/directory"
}

provider "acme" {
  alias = "staging"
  server_url = "https://acme-staging-v02.api.letsencrypt.org/directory"
}
```

**Comment by vancluever** [comment]
@uovobw lego generally bundles the CA (intermediate) certificate after the server certificate, so the provider expects the cert to be the first certificate in the bundle during examination. I was testing the provider yesterday with no issues, and [the current lego bundling code](https://github.com/go-acme/lego/blob/3d13faf68920543a393ad6cdfdea429627af2d34/acme/api/certificate.go#L40-L44)  does seems to indicate that this has not changed.

Can you tell me if:

* This certificate has been renewed recently
* Has the state been manipulated in any way?
* Have you tried downgrading the provider and trying again (set a [version constraint](https://www.terraform.io/docs/configuration-0-11/providers.html#provider-versions) in your Terraform config to do so.)
* Can you validate the certificate in `certificate_pem`:
  * Obtain this from your current state.
  * Save it to a file, ie: `certificate.crt` (you will have to convert the escaped newlines to real ones)
  * Run the following command. You will need OpenSSL installed:

```
openssl x509 -text -in certificate.crt
```

OpenSSL should then output the certificate. Verify that the subject line contains your domain name (same as `common_name`) and not the Let's Encrypt CA.

Thanks!

**Comment by uovobw** [comment]
thanks for the quick reply.
i'll reply point by point:
- the certificate had not been renewed recently, there was more than 1 month of time left
- the state had not been manipulated except than by terraform itself, and only for other different stuff (that is why i do not understand the reason for it failure now)
- i have tried setting the `version` string in the provider configuration from `>= 1.1.0, <=1.2.0` thus testing with `1.0.0`, `1.1.0` and `1.2.0`, no change in behaviour
- i have tested the certificate (i actually loaded in `webfsd` but the mechanism is the same) and it was working

i had to get the thing working to unblock some teammates - we are under a time constraint - and now i have solved the issue by defining a number of `file("/path/to/chain.pem")` templates in the terraform configuration as follows:
```
resource "aws_acm_certificate" "cc_ssl_certificate" {
  private_key = "${file("./privkey1.pem")}"
  certificate_body = "${file("./cert1.pem")}"
  certificate_chain = "${file("./chain1.pem")}"
}
```
with the files being generated by an ad-hoc `certbot` invocation together with the `certbot-dns-route53` plugin.
let me know how/if i can help debug the problem further.

**Comment by vancluever** [comment]
@uovobw I'm definitely still curious what the contents of `certificate_pem` were for you. Let me know what comes of that.

**Comment by uovobw** [comment]
i have managed to download the previous state file and inspect it, in the `certificate_pem` value there was something like what looked like a shasum of something, i was not able to find the "old" value.

**Comment by uovobw** [comment]
i have removed the registration and tried to create a new one in the state file - by running it with a different local file - but i get the same error upon creation. i have downloaded and built the go code for the `terraform-provider-acme` binary and replaced it in the `.terraform` directory as specified by the doc but i annot seem to get rid of the error

**Comment by vancluever** [comment]
@uovobw PEM certificates are displayed as base64-encoded blocks, similar to below:

```
-----BEGIN CERTIFICATE-----
abASDFasDFWfawed23qdfwef24FQEFASDFASDFSAFASFDASF....
-----END CERTIFICATE-----
```

If that's what you're seeing, you've found the right data. That's what you want to use when displaying the certificate data with OpenSSL.

**Comment by uovobw** [comment]
@vancluever that is what is there _now_ when i test the certificate, i went back in the terraform state file versions - thanks aws for versioning bucket objects - and during the period in the last couple of days in which i have been seeing the error there seems to only be this short string. that is a very very strange behaviour that i do not explain.

**Comment by brosner** [comment]
I am seeing this error too. One thing I'll add is that before I starting hitting this error, terraform notified me of a change:

```
certificate_pem:    "-----BEGIN CERTIFICATE-----<cert>-----END CERTIFICATE-----\n" => "<computed>"
```

**Comment by brosner** [comment]
My state file at the time of this error doesn't have a `certificate_pem` key under the attributes for `acme_certificate.certificate`.

**Comment by vancluever** [comment]
Thanks for the extra information @brosner. 

Can you give any any additional information that you have:

* What Terraform version are you using? Did you update to the latest RC recently?
* Have you updated the provider to the latest version (I don't recommend you do at this time, just curious).

**Comment by vancluever** [comment]
Just FYI, I was unable to reproduce this by (using Terraform 0.11.14):

* Create a certificate
* Upgrade to 1.3.0 and set `max_days_remaining` to 720 days to force a certificate renewal
* Set `max_days_remaining` back to its initial value

Any more information that can be provided would be appreciated!

**Comment by uovobw** [comment]
@vancluever i have fixed the issue as explained above. when (if) the problem will present itself again i'll comment on this ticket. i am pretty sure it must be something to do with the renewal process  in itself because now that the cert is not to be renewed the process completes flawlessly. thanks for the investigation

**Comment by DavidBennettUK** [comment]
I've also had this issue today. Using Terraform v0.11.13 and 1.3.0 of this provider, using `gcloud` for the `dns_challenge`

I ran Terraform today, which triggered the renewal for a certificate. It planned to update:
`acme_certificate.default certificate_pem: "XXXXXXXX" => <computed>`

This failed with the same issue as #51 :
```
* acme_certificate.default: acme: Error -> One or more domains had a problem:
acme: error presenting token: 2 errors occurred:
        * googlecloud: no public zone found for XXXXXXXX.
        * error encountered while presenting token for DNS challenge: googlecloud: no public zone found for domain XXXXXXXX.
```

Running Terraform again resulted in the plan failing with `unable to calculate time to certificate expiry: First certificate is a CA certificate`

Afterwards, I tried tainting the certificate resource to force a recreate using v1.1.2 of the provider. This didn't fail with the 'no public zone' issue, but still had the 'First certificate is a CA certificate' issue.

I then tried v1.0.1 of the provider, without tainting to recreate, and Terraform ran successfully - the only action it took was deleting the deposed `acme_certificate` resource.

I hope some information there helps!

**Comment by vancluever** [comment]
Thanks all for the info. I have a reproduction now and am pretty sure I know what's wrong, namely a lack of partial state mode is causing the certificate to be dropped if renewal is pending and an error occurs.

I should have a fix in early next week, including a fix that also recovers the existing certificate so that any states that have been affected by this that have not yet been recovered should be corrected.

Thanks for your patience!

**Comment by vancluever** [comment]
Update:

#60 fixes the issue so that the certificate is not dropped during update.

#59 is already in which will fix anybody affected by this that just haven't re-created their certificates.

As mentioned last night these should be released as 1.3.2 early next week!

**Comment by uovobw** [comment]
@vancluever there seems to be a problem with some internal tagging, the new `1.3.2` provider is marked as incompatible with terraform `0.12`:

```
Initializing provider plugins...
- Checking for available provider plugins...
- Downloading plugin for provider "aws" (terraform-providers/aws) 2.12.0...

Provider "acme" v1.3.2 is not compatible with Terraform 0.12.0.

Provider version 1.2.1 is the latest compatible version. Select it with 
the following constraint:

    version = "~> 1.2"

Terraform checked all of the plugin versions matching the given constraint:
    (any version)

Consult the documentation for this provider for more information on
compatibility between provider and Terraform versions.

Alternatively, upgrade to the latest version of Terraform for compatibility with newer provider releases.


```

**Comment by uovobw** [comment]
something strange has happened because i have run the `terraform init -upgrade` command that gave the error above. i now have the following terraform versions:
```
Terraform v0.12.0
+ provider.acme v1.3.1
+ provider.aws v2.12.0
+ provider.template v2.1.2
+ provider.tls v2.0.1
```
but the certificate error has disappeared and i can manipulate the acme certificates with no issues. i'll wait your input before doing any more tests as it's now working and i don't want to break the spell before i finish some pending stuff. thanks a lot for fixing this in the meantime!

**Comment by vancluever** [comment]
@uovobw sorry, the 1.3.2 provider was accidentally deployed for Terraform 0.11 only. I’ll be working on deploying 1.3.3 properly for Terraform 0.12 ASAP.

**Comment by uovobw** [comment]
@vancluever no problem at all, i wrote just to make sure it was a known issue! thanks for the info and for the great effort.

**Comment by vancluever** [comment]
@uovobw 1.3.3 is out now and should work with Terraform 0.12. Let me know how it goes!

**Comment by vancluever** [comment]
Closing this one out, assuming fixed, submit a new issue if it's still a problem!
